lib/tree.rb

```diff
@@ -16,51 +18,264 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log n)
+  # >> Assuming balanced BST, O(log n) time complexity (half the num nodes to check each time)
+  # >> If not balanced, could be O(n)
+  # Space Complexity: O(1)
+  # >> keep track of current tree node
+  def add(key, value = nil)
+    return @root = TreeNode.new(key, value) if @root.nil?
+    
+    current = @root
+    until current.nil?
+      if key > current.key
+        if !current.right
+          current.right = TreeNode.new(key, value)
+          return
+        else
+          current = current.right
+        end
+      else
+        if !current.left
+          current.left = TreeNode.new(key, value)
+          return
+        else
+          current = current.left
+        end
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # >> Assuming balanced BST, O(log n) time complexity (half the num nodes to check each time)
+  # >> If not balanced, could be O(n)
+  # Space Complexity: O(1)
+  # >> Keep track of current node
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+
+    current = @root
+    until current.nil?
+      return current.value if current.key == key
+
+      current = if key > current.key
+                  current.right
+                else
+                  current.left
+                end
+    end
+
+    return nil
+  end
+
+  def find_return_node(key)
+    return nil if @root.nil?
+
+    current = @root
+    until current.nil?
+      return current if current.key == key
+
+      current = if key > current.key
+                  current.right
+                else
+                  current.left
+                end
+    end
+
+    return nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # >> hit every node in tree during traversal
+  # Space Complexity: O(log n)
+  # >> go down one side of tree at a time --> stack has height # of levels
+  # >> O(n) if unbalanced
   def inorder
```

Comment:
👍   However, this is O(n) for both space and time.

Reply:
Ah -- forget that we are storing and returning all the visited nodes!